core/store/store.go

```diff
@@ -0,0 +1,37 @@
+package store
+
+import dbm "github.com/tendermint/tm-db"
+
+// KVStore describes the basic interface for interacting with key-value stores.
+type KVStore interface {
+
+	// Get returns nil iff key doesn't exist. Panics on nil key.
+	Get(key []byte) []byte
+
+	// Has checks if a key exists. Panics on nil key.
+	Has(key []byte) bool
+
+	// Set sets the key. Panics on nil key or value.
+	Set(key, value []byte)
+
+	// Delete deletes the key. Panics on nil key.
+	Delete(key []byte)
+
+	// Iterator iterates over a domain of keys in ascending order. End is exclusive.
+	// Start must be less than end, or the Iterator is invalid.
+	// Iterator must be closed by caller.
+	// To iterate over entire domain, use store.Iterator(nil, nil)
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// Exceptionally allowed for cachekv.Store, safe to write in the modules.
+	Iterator(start, end []byte) Iterator
+
+	// ReverseIterator iterates over a domain of keys in descending order. End is exclusive.
+	// Start must be less than end, or the Iterator is invalid.
+	// Iterator must be closed by caller.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// Exceptionally allowed for cachekv.Store, safe to write in the modules.
+	ReverseIterator(start, end []byte) Iterator
+}
+
+// Iterator is an alias db's Iterator for convenience.
+type Iterator = dbm.Iterator
```

Comment:
Did you consider redefining `tm-db/Iterator` in core to break the dependency on tendermint?

Reply:
```suggestion
type Iterator interface {
	// Domain returns the start (inclusive) and end (exclusive) limits of the iterator.
	// CONTRACT: start, end readonly []byte
	Domain() (start []byte, end []byte)

	// Valid returns whether the current iterator is valid. Once invalid, the Iterator remains
	// invalid forever.
	Valid() bool

	// Next moves the iterator to the next key in the database, as defined by order of iteration.
	// If Valid returns false, this method will panic.
	Next()

	// Key returns the key at the current position. Panics if the iterator is invalid.
	// CONTRACT: key readonly []byte
	Key() (key []byte)

	// Value returns the value at the current position. Panics if the iterator is invalid.
	// CONTRACT: value readonly []byte
	Value() (value []byte)

	// Error returns the last error encountered by the iterator, if any.
	Error() error

	// Close closes the iterator, relasing any allocated resources.
	Close() error
}
```

Reply:
It would make this incompatible with the existing interfaces

Reply:
I think for now it's fine as is, but maybe we exclude it from the service interface as it might break in the near future. tm-db -> cosmos-db and possibly other changes.

Reply:
Well we do need it for collections and orm

Reply:
We can make changes before we tag core 1.0

Reply:
I don't see too many usages of the typed iterator as a return param, am I searching wrong?

![image](https://user-images.githubusercontent.com/1236532/207961784-dd9bc3c7-c13f-4ca1-99e8-bb902998bce5.png)